books/kestrel/c/transformation/split-fn.lisp

```diff
@@ -0,0 +1,444 @@
+
+(include-book "centaur/fty/deftypes" :dir :system)
+(include-book "kestrel/std/util/defirrelevant" :dir :system)
+(include-book "kestrel/std/util/error-value-tuples" :dir :system)
+
+(include-book "../syntax/abstract-syntax-operations")
+(include-book "deftrans")
+(include-book "utilities/free-vars")
+
+(local (include-book "kestrel/built-ins/disable" :dir :system))
+(local (acl2::disable-most-builtin-logic-defuns))
+(local (acl2::disable-builtin-rewrite-rules-for-defaults))
+(set-induction-depth-limit 0)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defxdoc+ split-fn
+  :parents (transformation-tools)
+  :short "A C-to-C transformation to split a function in two."
+  :long
+  (xdoc::topstring
+    (xdoc::p
+      "This transformation takes the identifier of the function it is to split,
+       the name of the new function to be generated, and the location of the
+       split, represented as a natural number corresponding to the number of
+       statements in the function body before the split.")
+    (xdoc::p
+      "This transformation is a work in progress, and may fail in certain
+       cases. For one, it does not currently recognize multiple variables
+       declared at once (e.g. @('int x, y;')). It may also fail on variables
+       which have been declared but not yet initialized at the split point."))
+  :order-subtopics t
+  :default-parent t)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defruled ident-listp-when-ident-setp
+  (implies (ident-setp set)
+           (ident-listp set))
+  :induct t
+  :enable ident-setp)
+
+(local (in-theory (enable ident-listp-when-ident-setp)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(encapsulate ()
+  (set-induction-depth-limit 1)
+
+  (fty::defomap ident-decl-map
+    :key-type ident
+    :val-type decl
+    :pred ident-decl-mapp))
+
+(defrulel ident-listp-of-strip-cars-when-ident-decl-mapp
+  (implies (ident-decl-mapp map)
+           (ident-listp (strip-cars map)))
+  :induct t
+  :enable (strip-cars
+           ident-decl-mapp))
+
+(defrulel decl-listp-of-strip-cdrs-when-ident-decl-mapp
+  (implies (ident-decl-mapp map)
+           (decl-listp (strip-cdrs map)))
+  :induct t
+  :enable (strip-cdrs
+           ident-decl-mapp))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define expr-ident-list
+  ((idents ident-listp))
+  :returns (exprs expr-listp)
+  (if (endp idents)
+      nil
+    (cons (expr-ident (first idents))
+          (expr-ident-list (rest idents)))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define decl-to-paramdecl
+  ((decl declp))
+  :returns (mv erp
+               (paramdecl paramdeclp))
+  (b* (((reterr) (c$::irr-paramdecl)))
+    (decl-case
+      decl
+      :decl (b* (((when (endp decl.init))
+                  (reterr t))
+                 ((initdeclor initdeclor) (first decl.init)))
```

Comment:
Since it's only using `(first decl.init)`, I suppose that there may be an assumption that the `decl` passed as input should declare just one identifier? It would be good to document this kind of things.